core/kernel/tpm.c

```diff
@@ -0,0 +1,154 @@
+
+static vaddr_t tpm_log_addr;
+static size_t tpm_log_size;
+
+/*
+ * Check whether the node at @offs contains TPM Event Log information
+ * (tpm_event_log_sm_addr and tpm_event_log_size properties) or not.
+ *
+ * @offs is the offset of the node that describes the device in @fdt.
+ * @buf will contain the phy address of the TPM Event log.
+ * @size will contain the size of the mapped area.
+ *
+ * Returns the size of the mapped area or @DT_INFO_INVALID_TPM_LOG upon failure
+ */
+#ifdef CFG_DT
+static int read_dt_tpm_log_info(void *fdt, int node, paddr_t *buf, size_t *size)
+{
+	const uint32_t *property = NULL;
+	uint64_t zero_addr;
+	int len_prop;
+	paddr_t log_addr;
+	int err;
+
+	/*
+	 * Get the TPM Log address.
+	 */
+	property = fdt_getprop(fdt, node, "tpm_event_log_sm_addr", &len_prop);
+
+	if (!property  || len_prop != sizeof(uint32_t) * 2)
+		return DT_INFO_INVALID_TPM_LOG;
+
+	log_addr = fdt32_to_cpu(property[1]);
+
+	zero_addr = 0;
+	err = fdt_setprop(fdt, node, "tpm_event_log_sm_addr", &zero_addr,
+								sizeof(uint32_t) * 2);
+	if (err < 0) {
+		EMSG("Error setting property \'tpm_event_log_sm_addr\' to zero\n");
+		return err;
+	}
+
+	/*
+	 * Get the TPM Log size.
+	 */
+	property = fdt_getprop(fdt, node, "tpm_event_log_size", &len_prop);
+
+	if (!property || len_prop != sizeof(uint32_t))
+		return DT_INFO_INVALID_TPM_LOG;
+
+	*size = fdt32_to_cpu(property[0]);
+	*buf = log_addr;
+
+	return *size;
+}
+#endif
+
+static void get_tpm_phys_params(void *fdt, paddr_t *addr, size_t *size)
+{
+#ifdef CFG_DT
+	int node;
+	const char *dt_tpm_match_table = {
+		"arm,nt_fw",
+	};
+
+	if (!fdt) {
+		EMSG("TPM: No DTB found");
+		return;
+	}
+
+	node =
+	    fdt_node_offset_by_compatible((const void *) fdt, -1,
+					  dt_tpm_match_table);
+
+	if (node < 0) {
+		EMSG("TPM: Fail to find TPM node %i", node);
+		return;
+	}
+
+	if (read_dt_tpm_log_info
+	    ((void *) fdt, node, addr, size) == DT_INFO_INVALID_TPM_LOG) {
+		EMSG("TPM: Fail to retrieve DTB properties from node %i",
+		     node);
+		return;
+	}
+#else
+	(void) fdt;
+	*size = CFG_TPM_MAX_LOG_SIZE;
+	*addr = CFG_TPM_LOG_BASE_ADDR;
+#endif				/* CFG_DT */
+}
+
+TEE_Result get_tpm_event_log(void *buf, size_t *size)
+{
+	if (!buf) {
+		/*
+		 * Sanity check in case an uninitialized buffer is passed.
+		 */
+		EMSG("TPM: Invalid buffer");
+		return -1;
+	}
+
+	if (*size < tpm_log_size) {
+		EMSG("TPM: Not enough space for the log: %zu, %lu",
+		     *size, tpm_log_size);
+		return -1;
+	}
+
+	*size = tpm_log_size;
+	memcpy(buf, (void *) tpm_log_addr, tpm_log_size);
+
+	return 0;
+}
+
+void map_tpm_log_area(void *fdt)
+{
+	paddr_t log_addr;
+	unsigned int paging_size;
+
+	log_addr = 0;
+	get_tpm_phys_params(fdt, &log_addr, &tpm_log_size);
+
+	IMSG("TPM Event log PA: %p", (void *) log_addr);
+	IMSG("TPM Event log size: %zu Bytes", tpm_log_size);
+
+	paging_size = ROUNDUP(tpm_log_size, SMALL_PAGE_SIZE);
+
+	if (!core_mmu_add_mapping
+	    (MEM_AREA_RAM_SEC, log_addr, paging_size)) {
```

Comment:
This is a strange place to break the line at, please fold up as much as possible.